### DISCUSSION TITLE
raster tessellation to 3D array?

### DISCUSSION THREAD

**Discussion by xshang93** [neperfepx/neper#486] [Q&A] [answered]
Hi, I wonder if there is a way to convert `.tesr` files (or any other formats that describe a raster tessellation) to a 3D array containing grain orientations? Say if I have a 32x32x32 raster, how to generate an array of 32x32x32x3 to describe it? The last 3 channels are 3 Euler angles, for instance.

Thank you!

**Comment by rquey** [comment, marked as answer]
It is possible using `-statvox euler-bunge`. It will write the voxel orientations on successive lines. *x* varies first, *y* second and *z* last (like in the tesr itself).

Needs version `4.4.2-17`.

Example:
```
$ neper -T -n 2 -for tesr -tesrsize 3 -statvox euler-bunge

========================    N   e   p   e   r    =======================
Info   : A software package for polycrystal generation and meshing.
Info   : Version 4.4.2-17
Info   : Built with: gsl|muparser|opengjk|openmp|nlopt|libscotch (full)
Info   : Running on 8 threads.
Info   : <https://neper.info>
Info   : Copyright (C) 2003-2022, and GNU GPL'd, by Romain Quey.
Info   : Loading initialization file `/home/rquey/.neperrc'...
Info   : ---------------------------------------------------------------
Info   : MODULE  -T loaded with arguments:
Info   : [ini file] none
Info   : [com line] -n 2 -for tesr -tesrsize 3 -statvox euler-bunge
Info   : ---------------------------------------------------------------
Info   : Reading input data...
Info   : Creating domain...
Info   : Creating tessellation...
Info   :   - Setting seeds... 
Info   :   - Generating crystal orientations...
Info   :   - Running tessellation...
Info   : Rasterizing tessellation...
Info   : Writing results...
Info   :     [o] Writing file `n2-id1.tesr'...
Info   :     [o] Wrote file `n2-id1.tesr'.
Info   : Writing statistics...
Info   :     [o] Writing file `n2-id1.stvox'...
Info   :     [o] Wrote file `n2-id1.stvox'.
Info   : Elapsed time: 0.013 secs.
========================================================================

$ more n2-id1.stvox 
41.813165532991 78.438807181793 -75.984842545419
236.335386748008 80.072535051284 -95.270156451555
236.335386748008 80.072535051284 -95.270156451555
236.335386748008 80.072535051284 -95.270156451555
236.335386748008 80.072535051284 -95.270156451555
236.335386748008 80.072535051284 -95.270156451555
236.335386748008 80.072535051284 -95.270156451555
236.335386748008 80.072535051284 -95.270156451555
236.335386748008 80.072535051284 -95.270156451555
41.813165532991 78.438807181793 -75.984842545419
41.813165532991 78.438807181793 -75.984842545419
41.813165532991 78.438807181793 -75.984842545419
41.813165532991 78.438807181793 -75.984842545419
41.813165532991 78.438807181793 -75.984842545419
236.335386748008 80.072535051284 -95.270156451555
41.813165532991 78.438807181793 -75.984842545419
236.335386748008 80.072535051284 -95.270156451555
236.335386748008 80.072535051284 -95.270156451555
41.813165532991 78.438807181793 -75.984842545419
41.813165532991 78.438807181793 -75.984842545419
41.813165532991 78.438807181793 -75.984842545419
41.813165532991 78.438807181793 -75.984842545419
41.813165532991 78.438807181793 -75.984842545419
41.813165532991 78.438807181793 -75.984842545419
41.813165532991 78.438807181793 -75.984842545419
41.813165532991 78.438807181793 -75.984842545419
41.813165532991 78.438807181793 -75.984842545419
```

**Reply by xshang93** [reply]
Awesome. Thanks @rquey . I'm still on 4.2.1 though, but I wrote some lines to do that in Python.

A not very relevant question, how does one generate mapped meshes from `.tesr` files, and can they be used for simulations in fepx? The documentation says mapped meshes can be generated by I didn't find how. If I just run `$ neper -M file_name.tesr`I just get `Free meshing not available for 3D tesrs.`, so it alway tries to free mesh the whole thing.

**Reply by rquey** [reply]
You have to use `-elt hex`, but FEPX does not take hex meshes.

**Reply by xshang93** [reply]
> You have to use `-elt hex`, but FEPX does not take hex meshes.

ok sounds good. Thanks!

**Reply by xshang93** [reply]
Hi @rquey , I wanted to follow up on this one. Is it still the case that neper can't free mesh rasterized structures, and fepx doesn't take hex mesh? Is there any other way you may know that could solve a rasterized microstructure generated in Neper?

Thanks!

**Reply by rquey** [reply]
FEPX does not take hex meshes and it won't happen anytime soon.

Neper can mesh rasterized structures in tet elements, but only after some approximation. See https://neper.info/papers.html#id2 for the theory and https://neper.info/doc/tutorials/morpho_tesr_mesh.html for an application.

Example:
```
# Generate sample tesr
neper -T -n 10 -for tesr -o px

# Approximate by tess
neper -T -n 10 -morpho "tesr:file(px.tesr)" -o px -reg 1

# Mesh and visualize
neper -M px.tess
neper -V px.tesr -print img1
neper -V px.tess,px.msh -print img2
```
![image](https://github.com/neperfepx/neper/assets/2659376/97ed20fb-4ad0-40ef-91aa-5cf397539ff1)
![image](https://github.com/neperfepx/neper/assets/2659376/9f1152d2-2a13-4637-82ca-03a08372166a)
